corehq/apps/domain/tests/test_login_and_domain_required.py

```diff
@@ -0,0 +1,149 @@
+
+        self.domain_patcher = mock.patch('corehq.apps.domain.models.Domain.get_by_name', return_value=self.domain)
+        self.get_domain_by_name = self.domain_patcher.start()
+        self.addCleanup(self.domain_patcher.stop)
+
+    @staticmethod
+    def _create_domain():
+        domain = mock.Mock(Domain)
+        domain.is_active = True
+        domain.is_snapshot = False
+        return domain
+
+    @staticmethod
+    def _create_user():
+        user = mock.Mock()
+        user.is_superuser = False
+        user.is_member_of = mock.Mock(return_value=False)
+        return user
+
+    def _create_request(self, url, user):
+        request = self.factory.get(url)
+        request.user = request.couch_user = user
+        request.path = 'access_required.html'
+        request.domain = self.domain_name
+
+        return request
+
+    def process_request(self):
+        return process_request(self.request, self.domain_name)
+
+    def test_invalid_domain__displays_404(self):
+        self.get_domain_by_name.return_value = None
+
+        with self.assertRaises(Http404):
+            self.process_request()
+
+    # Reverse is stubbed here because actual reverse lookups take ~ 1 second
+    @mock.patch('corehq.apps.domain.decorators.reverse')
+    @mock.patch('corehq.apps.domain.decorators.messages', mock.Mock())
+    def test_inactive_domain__redirects_to_domain_selection(self, reverse):
+        self.domain.is_active = False
+        reverse.return_value = '/domain/select/'
+
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+        reverse.assert_called_with('domain_select')
+
+    def test_two_factor_required_but_missing__displays_alert(self):
+        self.domain.two_factor_auth = True
+        self.user.is_member_of.return_value = True
+        self.user.two_factor_disabled = False
+        self.user.is_verified = mock.Mock(return_value=False)
+
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_snapshot_domain_without_previewer__returns_not_found(self):
+        self.domain.is_snapshot = True
+        self.user.is_previewer = mock.Mock(return_value=False)
+        with self.assertRaises(Http404):
+            self.process_request()
+
+    def test_snapshot_domain_with_previewer__allows_access(self):
+        self.domain.is_snapshot = True
+        self.user.is_previewer = mock.Mock(return_value=True)
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_user_belongs_to_domain__returns_200(self):
+        self.user.is_member_of.return_value = True
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    @mock.patch('corehq.apps.domain.decorators.has_privilege', mock.Mock(return_value=False))
```

Comment:
You might want a narrower mock here - `has_privilege` is a very broadly used function.  This should return `False` when the privilege in question is the one about project access, but not otherwise.  Other privilege checks typically happen in a separate decorator, so maybe that's a moot point.

Reply:
Hmm, I don't generally like to explicitly specify arguments to mocks, because I feel it makes mocks even more verbose, and also couples the test to the implementation more. I can see the argument here, however, especially if there's a chance that other privileges might be checked in a future version of this code. If you think the code would benefit from a more explicit check, I'll add it in

Reply:
no strong preference, just wanted to flag